### ISSUE TITLE
Restore documentation

### ISSUE THREAD

**Issue by blinkiz** [zrepl/zrepl#458]
Hello
Am new to ZFS and also zrepl.
Feature request! One thing am missing is restore instructions. It does not exist any documentation how to restore without breaking something. There is no guide how to routinely test restoring data from a remote zrepl sink.
I would like for this to exist. 🙂

**Comment by problame** [comment]
Could you describe what the UX for restore should look like? I.e. CLI commands, config syntax, procedures, etc.
Some questions to get the ideas going:
- Scenario 1: Full system restore after I spilled coffee over my laptop. I got a new laptop with a blank hard drive, probably a live USB stick. What do I do.
- Scenario 2: Accidentally deleted some data that is not in local snapshots but in remote snapshots.
  - Variant 1: I accidentally destroyed a dataset that contains important user data.
  - Variant 2: I accidentally destroyed the dataset that contained the zrepl config.
  - Variant 3: I accidentally deleted a file and want to restore it from backup.

Don't feel bound by what you know about zrepl / ZFS, just give tutorial-level steps on how you'd imagine the perfect tool to work.

**Comment by blinkiz** [comment]
Thanks for asking. Okay, lets see..

For restoring I would like to use jobs names defined in configuration. That is easy to understand.
Shell auto-completions should exist for both SRC and DEST so it is easy to TAB to right snapshot.


```
Restore from defined jobs in configuration. 

Usage:
  zrepl restore [flags] SRC DEST
  
Examples:  
  zrepl restore sink::storage/zrepl/sink/var/db@zrepl_20060102_150405_000 prod_to_backups::zroot/var/db2
  Restores from job "sink" into job "prod_to_backups" with path zroot/var/db2 
  that may or may not already exist.
  
  zrepl restore prod_to_backups::storage/zrepl/sink/var/db@zrepl_20060102_150405_000 prod_to_backups::zroot/var/db
  Restore existing snapshot within same job definition.

Flags:
  -h, --help   help for restore
  -f, --force  Overwrite if destination already exist without first asking

Global Flags:
      --config string   config file path
```

With this structure I do not need to know how ZFS works when restoring. I do not need to understand how the transport works because that is taken care of within job definitions. This is a good addition to manually using ZFS tools.

What do you think?

**Comment by problame** [comment]
I like that you presented the a fully fledged CLI help output :)

> zrepl restore sink::storage/zrepl/sink/var/db@zrepl_20060102_150405_000 prod_to_backups::zroot/var/db2
  Restores from job "sink" into job "prod_to_backups" with path zroot/var/db2 
  that may or may not already exist.

I assume this command is to be run on the `push` side? How does the push side know the job name on the sink side? (assuming the sink side job is named `sink`).

> zrepl restore prod_to_backups::storage/zrepl/sink/var/db@zrepl_20060102_150405_000 prod_to_backups::zroot/var/db
  Restore existing snapshot within same job definition.

I don't underrstand why it's `prod_to_backups::storage/zrepl/sink/var/db@zrepl_20060102_150405_000` and not `prod_to_backups::var/db@zrepl_20060102_150405_000` .

**Comment by blinkiz** [comment]
Hi there
> I assume this command is to be run on the `push` side? How does the push side know the job name on the sink side? (assuming the sink side job is named `sink`).

To make it easier all SRC or DEST targets needs to be in zrepl configuration file. If the sink side should push something back to production server then it needs to be some kind of passive job defined in configuration like 'prod_to_backups' I have used in my examples. My goal is start write `zrepl restore` and then TAB away in auto completion.

> I don't underrstand why it's `prod_to_backups::storage/zrepl/sink/var/db@zrepl_20060102_150405_000` and not `prod_to_backups::var/db@zrepl_20060102_150405_000` .

Yeah it should be `prod_to_backups::zroot/var/db@zrepl_20060102_150405_000`, I wrote wrong. I will update my example.

**Comment by problame** [comment]
> `zrepl restore sink::storage/zrepl/sink/var/db@zrepl_20060102_150405_000 prod_to_backups::zroot/var/db2`

ATM the receice-side paths are not part of the replication protocol.
Maybe we'll have to change this to support arbitrary dataset mappings ( #253 ) but ATM the `sink` does the prepent/prefixremove of `root_fs` completely transparent to the sender. I don't think we should change that.

> zrepl restore prod_to_backups::zroot/var/db@zrepl_20060102_150405_000 prod_to_backups::zroot/var/db

* Q1: So the idea here is that `zroot/var/db@zrepl_20060102_150405_000` is a locally present snapshot?
* Q2: If we do this restore we effectively rollback `zroot/var/db`. What happens to
  * the local snapshots that are newer than `@zrepl_20060102_150405_000`? Is that what the `--force` flag is about?
  * the remote snapshots that are newer than `@zrepl_20060102_150405_000`?
    * Option 1: Delete them, maybe gated by `--force` or `--rollback-remote`
    * Option 2: Let conflict handling do its work ( #408  )

Other issues:

* How do I, as a user, or say the autocompletion handler, learn about the available snapshots to be restored, ideally, the entire `sink::storage/zrepl/sink/var/db@zrepl_20060102_150405_000` line?

**Comment by blinkiz** [comment]
Interesting questions you have.
My viewpoint is from a user UX perspective. I do understand that many technical challenges exist. All backup solutions that do not encourage user to practice restoring data should be avoided. zrepl do not have a restore feature.

When I opened this issue 12 days ago I had the impression that only documentation was missing how to restore snapshots created with zrepl. I thought adding some documentation around zfs send/recv commands and it would be enough. After I while I realized that the feature did not exist and restoring was more complicated because of the state (hold) zrepl creates for snapshots. Am sorry but I will move on to something else for my needs. I do hope someone else will will help @problame to create this feature. He is doing a great job creating open source software. Please contribute by giving ideas how this restore feature should work. 🙂

**Comment by problame** [comment]
>  After I while I realized that the feature did not exist and restoring was more complicated because of the state (hold) zrepl creates for snapshots.

Just for the record: `zrepl zfs-abstraction [list|release-stale|release-all]` would probably have simplified things for you.

**Comment by bartmichu** [comment]
Not that a long time ago I needed to restore just a couple of files from my USB backup. I found myself scratching my head at first - how do I do that? And in a way that will not break zrepl jobs in the future?

Zrepl really is so easy to use and abstracts so much away that one doesn't need to know much about ZFS. Getting data back can be problematic, especially in a situation when backup is needed.. :)

It is great that you are thinking about the ```restore``` functionality, but in the meantime I would like to back the initial request for some hints in the documentation.

**Comment by problame** [comment]
> Not that a long time ago I needed to restore just a couple of files from my USB backup. I found myself scratching my head at first - how do I do that? And in a way that will not break zrepl jobs in the future?

What did you end up doing, in the end?

I think there are three major ways of doing restore ATM:
- Manually use `zfs rollback` on the production machine
- Manually do `zfs send backuppool/data/set/path@zrepl_... | zfs recv production/...`
- Clone or mount the snapshot on the backuppool, extract the relevant files, copy them back to the production pool.

> It is great that you are thinking about the restore functionality, but in the meantime I would like to back the initial request for some hints in the documentation.

Turning this into a separate issue, there might be some opportunity to work on docs at the end of the week.

**Comment by bartmichu** [comment]
@problame I went with your third option: mounting snapshot and manually restoring selected files. Nothing zrepl specific as far as I remember, but in a stressful situation it's good to have a list of specific steps. Also I didn't know if and how it is going to break future zrepl jobs.

**Comment by dionysius** [comment]
Can we give this whole thing some traction? **A backup system that can't restore is just broken** *

_* The word "can't" might be a bit exaggerated, but feels right to me as there is **zero** documented zrepl or lower level zfs instructions by use case. Even worse, some variants shouldn't be done since it can **break** your backup (as commented above) which should also be documented_

**Comment by jficz** [comment]
This is a selling (or rather _unselling_) point for me. Implementing zrepl as my backup backend but I have a hard time explaining how the _restore_ part is gonna work to the stakeholders. I think this could be mostly solved by proper documentation as the restore mechanics are inherently present in ZFS.

**Comment by midzelis** [comment]
Just discovered this issue after I created this discussion topic: https://github.com/zrepl/zrepl/discussions/730

I also think restore is vital, and at minimum, we need some instructions on how to manually restore using base ZFS commands, but ideally, having some built-in zrepl commands to make it easier would be best. 

Especially for the untrusted remote/encrypted ZFS scenarios - since that requires special consideration for managing/backing up the encryption keys.

**Comment by jficz** [comment]
> Especially for the untrusted remote/encrypted ZFS scenarios - since that requires special consideration for managing/backing up the encryption keys.

That is imho exactly why encrypted scenarios should _not_ fall under zrepl scope. Key management is usually implementation-specific and zrepl shouldn't make any assumptions about it. Encrypted backups should be handled transparently by zrepl by raw streams and key management/unlocking should fall outside zrepl scope completely.

That being said, some good practice examples can never hurt. I'm sure the community and developers have various tested and proven ways to tackle this problem.

**Comment by simseb** [comment]
I think recovery scenarios should not fall under zrepl implementation, but some examples in the documentation would be nice:

- How to recover specific files by mounting the snapshot
- How to recover a complete dataset via ssh using zfs send and receive between the backup server and the workstation
- etc...

**Comment by jficz** [comment]
well, as they say in the docs:

> zrepl is a one-stop, integrated solution for ZFS *replication*.

It doesn't say "backup" but imho backups are the most common use case for what `zrepl` provides.

`zrepl` does replication nicely, with all the bells and whistles, resumes, safeguards, etc. Technically a restore (on dataset/snapshot level) is "just" a reverse operation and I think it could be implemented as part of `zrepl`. However, I suspect such operations would be far less common in practice and they will likely be one-off events. Proper documentation of "what `zrepl` would do under the hood if it was asked to do this, this and that" would have significant impact on its own even without it being implemented.

**Comment by dionysius** [comment]
It does say baclupThe github repo description and readme say:

> zrepl is a one-stop ZFS backup & replication solution.

**Comment by InsanePrawn** [comment]
[While I can not speak for @problame or zrepl as a project in this regard; my personal opinions are as follows:]
**First off: I'm all for someone writing these restore docs or even having restore functionality inside zrepl.**

However, I feel like I have to warn people: If you're not familiar with how ZFS send or mounting a datasets works, you should probably not yet be using ZFS for anything worth backing up, in my frank opinion.

IMO people need to read a ZFS primer of some form before using zrepl, as zrepl can not replace an understanding of the various underlying ZFS mechanisms and behaviours and tooling.

ZFS is also not free enough of sharp edges (the various edges with encryption, or plain old dedup for starters) that people should be blindly installing ZFS + zrepl and then think they're golden.
zrepl can and will not protect you from those issues.

I think the docs we write for this should warn people that they ought to familiarize themselves with ZFS a little first, and also to practice the recovery procedure before taking the system into production, obviously. :)
